invenio_rdm_records/collections/api.py

```diff
@@ -0,0 +1,346 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2024 CERN.
+#
+# Invenio-RDM is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+"""Collections programmatic API."""
+
+from types import ClassMethodDescriptorType
+
+from invenio_db import db
+from sqlalchemy import select
+from werkzeug.utils import cached_property
+
+from .models import Collection as CollectionModel
+from .models import CollectionTree as CollectionTreeModel
+
+
+class Collection:
```

Comment:
Are we sure that this api layer is really needed, when the data model is not a classic record with a json schema? Can't the models be directly used in the service layer (as done in `invenio-jobs` for example)?
This class looks like a wrapper of the model, and the various methods can be moved to the model class.

Reply:
On one hand, I agree this class feels like a wrapper to the model, and it could be richer (e.g. I and Alex discussed storing the relation between collections in memory and then they could be "navigated" in memory). 

However, some ideas that made me implement a public API for collections:

1. Abstract some concepts about how we store collections in DB (materialized path). In my opinion, for the developer, things like `path` should be abstracted and encapsulated under a "nice looking" API.
2. Related to the one above, adding some public-facing utilities to improve the usability of collections, e.g. `collection.ancestors`, `collection.children`, create a collection based on a parent, resolving a collection by `id` or `slug, community_id` pair. 
3. `dump` / `to_dict` methods for serialization.

Reply:
My 2 cents

I understand your reasoning. At the same time, with this approach, we will end up having a wrapper classes that have most of the attributes and methods duplicated from the encapsulated model. And then we add a couple of extra methods.
I would not know what is the separation of concerns or abstraction, and if I have to add a new feature (a new query), I would not know if I have to put in the API or in the Model. What is the criteria that tells me where to put it?

For the `path`, maybe, I don't see it for the moment in the class.
For your second point, it can be done in the model, I don't see the difference.

A model is already an API class. What has been done for records is actually very different and in that case very much needed: the model is loaded into a `dict` object, with many other extra methods.

If we start creating wrappers everywhere, we will end up having more complex code, and less DRY. As a comparison, it looks a little bit like the props drilling problem in React.

As an alternative, if you really feel like that we need to separate things, then in this case I would go with inheritance instead of composition.

To recap: I think that our final objective is to have the simplest possible code. If the majority of the APIs exposed by a model are intended to be used, I would use the model directly or create inheritance.
Composition and wrapping make sense to me when you want to hide the majority of the features provided by the wrapped object.

Reply:
Those are excellent points, I will keep them in mind in the next PR. 

Thanks for the detailed explanation 🚀